pkg/job/discovery.go

```diff
@@ -61,33 +64,46 @@ func runDiscoveryJob(
 
 	maxMetricCount := metricsPerQuery
 	length := getMetricDataInputLength(job.Metrics)
-	partition := int(math.Ceil(float64(metricDataLength) / float64(maxMetricCount)))
-	logger.Debug("GetMetricData partitions", "total", partition)
+	partitionSize := int(math.Ceil(float64(metricDataLength) / float64(maxMetricCount)))
+	logger.Debug("GetMetricData partitions", "size", partitionSize)
 
-	var wg sync.WaitGroup
-	wg.Add(partition)
+	g, gCtx := errgroup.WithContext(ctx)
+	g.SetLimit(concurrencyLimit)
 
-	getMetricDataOutput := make([][]*cloudwatch.MetricDataResult, partition)
-	count := 0
+	mu := sync.Mutex{}
+	getMetricDataOutput := make([][]*cloudwatch.MetricDataResult, 0, partitionSize)
 
+	count := 0
 	for i := 0; i < metricDataLength; i += maxMetricCount {
-		go func(i, n int) {
-			defer wg.Done()
-			end := i + maxMetricCount
-			if end > metricDataLength {
-				end = metricDataLength
-			}
-			input := getMetricDatas[i:end]
-			data := clientCloudwatch.GetMetricData(ctx, logger, input, svc.Namespace, length, job.Delay, job.RoundingPeriod)
+		start := i
+		end := i + maxMetricCount
+		if end > metricDataLength {
+			end = metricDataLength
+		}
+		partitionNum := count
+		count++
+
+		g.Go(func() error {
+			logger.Debug("GetMetricData partition", "start", start, "end", end, "partitionNum", partitionNum)
+
+			input := getMetricDatas[start:end]
+			data := clientCloudwatch.GetMetricData(gCtx, logger, input, svc.Namespace, length, job.Delay, job.RoundingPeriod)
 			if data != nil {
-				getMetricDataOutput[n] = data
+				mu.Lock()
+				getMetricDataOutput = append(getMetricDataOutput, data)
+				mu.Unlock()
```

Comment:
Any reason to move to locking now vs index based? It's probably safer/easier to reason about regardless just wondering

Reply:
Yes and another reason was to remove the preallocation of the slice (size of `partitionSize`).